In [2]:
import pandas as pd

In [3]:
df = pd.read_csv('DataSets/ItemDataWithAiDescription.csv', index_col=0)

In [4]:
df

ProductName          Brand  Price  \
0           1017 ALYX 9SM Circle Logo Popover Hoodie  1017-alyx-9sm  325.0   
0                   1017 ALYX 9SM Side Zip Camo Pant  1017-alyx-9sm  443.0   
0            1017 ALYX 9SM x Blackmeans Denim Jacket  1017-alyx-9sm  779.0   
0   1017 ALYX 9SM Oversized Camo Nylon Bomber Jacket  1017-alyx-9sm  779.0   
0            1017 ALYX 9SM Nylon Logo Varsity Jacket  1017-alyx-9sm  489.0   
..                                               ...            ...    ...   
0                                YMC Embroidered Cap            ymc   62.0   
0                      YMC Ibiza '89 Pyramid T-Shirt            ymc   47.0   
0                      YMC Ibiza '89 Dancers T-Shirt            ymc   43.0   
0                       YMC Ibiza '89 Sunset T-Shirt            ymc   43.0   
0                                  Yogi Lennon Suede           yogi   95.0   

            Category Sizes                                    TextDescription  \
0   Hoodies & Sweats    []  Drawing influences from California skate cultu...   
0           Trousers    []  1017 ALYX 9SM innovative designs take cues fro...   
0    Coats & Jackets    []  Working together on a capsule that creates a r...   
0    Coats & Jackets    []  With an unmistakable experimental streetwear a...   
0    Coats & Jackets    []  Matthew Williams’s genre-blurring, future-orie...   
..               ...   ...                                                ...   
0        Accessories    []  You can never have too many caps, so add to yo...   
0           T-Shirts    []  YMC joins forces with Dave Swindells to celebr...   
0           T-Shirts    []  YMC continues to encourage us to create our ow...   
0           T-Shirts    []  Shot over a one week assignment for 20/20 maga...   
0           Footwear    []  One of Yogi’s  most recognised silhouettes, th...   

                               BulletPointDescription  \
0   ['100% Cotton', 'Drawstring Hood', 'Kangaroo P...   
0   ['100% Nylon', 'Elasticated Waist', 'Zip Closu...   
0   ['100% Cotton', 'Classic Collar', 'Button Clos...   
0   ['100% Nylon', 'Stand Up Collar', 'Snap Button...   
0   ['100% Polyester', 'Shawl Collar', 'Snap Butto...   
..                                                ...   
0   ['74% Cotton, 26% Linen', '8-Wale Corduroy', '...   
0   ['100% Organic Slub Cotton', 'Oversized Fit', ...   
0   ['100% Organic Slub Cotton', 'Oversized Fit', ...   
0   ['100% Organic Slub Cotton', 'Oversized Fit', ...   
0   ['Suede Uppers', 'Natural Rubber Outsoles', 'W...   

                                            ImageUrls  \
0   ['https://media.endclothing.com/media/f_auto,q...   
0   ['https://media.endclothing.com/media/f_auto,q...   
0   ['https://media.endclothing.com/media/f_auto,q...   
0   ['https://media.endclothing.com/media/f_auto,q...   
0   ['https://media.endclothing.com/media/f_auto,q...   
..                                                ...   
0   ['https://media.endclothing.com/media/f_auto,q...   
0   ['https://media.endclothing.com/media/f_auto,q...   
0   ['https://media.endclothing.com/media/f_auto,q...   
0   ['https://media.endclothing.com/media/f_auto,q...   
0   ['https://media.endclothing.com/media/f_auto,q...   

                                        AIDescription  
0   The item is a hoodie from the category of Hood...  
0   The item is categorized as trousers and is spe...  
0   The item is a denim jacket categorized under C...  
0   The item is an **Oversized Camo Nylon Bomber J...  
0   The item is a coat categorized as a varsity ja...  
..                                                ...  
0   The YMC Embroidered Cap falls under the catego...  
0   The YMC Ibiza '89 Pyramid T-Shirt falls under ...  
0   The YMC Ibiza '89 Dancers T-Shirt is categoriz...  
0   The YMC Ibiza '89 Sunset T-Shirt is categorize...  
0   The Yogi Lennon Suede is a pair of footwear ca...  

[14912 rows x 9 columns]

In [5]:
# Dictionar to clean the category
colour_options = [
    "black",
    "white",
    "green",
    "navy",
    "brown",
    "blue",
    "olive",
    "beige",
    "grey",
    "silver",
    "yellow",
    "orange",
    "red",
    "cream",
    "orange",
    "pink",
    "purple",
    "teal",
    "gold",
    "clear",
    "tan"
]

pattern_options = [
    "plain",
    "graphic",
    "printed",
    "striped",
    "check",
    "floral",
    "embroidered",
    "plaid",
    "ribbed",
    "jacquard",
    "geometric",
    "pleated",
    "solid",
    "camo",
    "abstract",
    "quilted",
    "patchwork",
    "monogram",
    "distressed"
]

fabric_options = [
    "cotton",
    "polyester",
    "nylon",
    "leather",
    "wool",
    "silk",
    "suede",
    "canvas",
    "polyamide",
    "acetate",
    "mix-cotton",
    "mix-polyester",
    "mix-nylon",
    "mix"
]

material = [
    "silver",
    "gold",
    "wood",
    "aluminium",
    "steel",
    "concrete",
    "plastic"
]

fit_options = [
    "relaxed",
    "oversized",
    "regular",
    "boxy",
    "loose",
    "slim",
    "straight leg",
    "wide leg",
    "tapered",
    "slouchy",
    "baggy",
    "classic",
    "wide",
    "tailored",
    "cropped",
    "tight",
    "skinny"
]

gender_audience_options = [
    "unisex",
    "men",
    "Unisex",
    "women",
    "children"
]

age_audience_options = [
    "adult",
    "teen",
    "children",
    "toddler",
    "baby"
]

season_options = [
    "summer",
    "autumn",
    "winter",
    "spring",
    "all seasons"
]

occasion_option = [
    "casual",
    "athletic",
    "smart casual",
    "formal",
    "travel",
    "outdoor",
    "hiking"
]

sleeve_length_options = [
    "long",
    "short",
    "sleeveless",
    "elbow length",
    "three-quarter"
]

neckline_options = [
    "hooded",
    "crew neck",
    "classic collar",
    "drawstring hood",
    "v-neck",
    "camp collar",
    "high neck",
    "turtle-neck",
    "high collar",
    "wide collar"
    "stand collar",
    "relaxed collar",
    "button down",
    "mock neck",
    "open collar",
    "spread collar",
    "shawl collar",
    "polo collar",
    "pointed collar",
    "full zip",
    "half zip",
    "three-quarter zip"
]

closure_type_options = [
    "drawstring",
    "button",
    "zip"
]

pocket_details_options = [
    "kangaroo pocket",
    "side pocket",
    "chest pocket",
    "multi pocket",
    "front pocket",
    "pouch pocket"
]

embellishments_options = [
    "print",
    "embroidery",
    "logo",
    "patch",
    "embossing",
    "graphic"
]

occasion_option = [
    "casual",
    "athletic",
    "smart casual",
    "formal",
    "travel",
    "outdoor",
    "festival",
    "everyday"
]

category_options = [
    "clothing",
    "footwear",
    "accessories",
    "lifestyle"
]

sub_category_options = {
    "clothing": [
        "coats & jackets",
        "hoodies & sweats",
        "sweat pants",
        "t-shirts",
        "jeans",
        "knitwear",
        "shirts",
        "trousers",
        "shorts",
        "polo shirts",
        "loungewear",
        "swimwear",
        "underwear"
    ],
    "footwear": [
        "sneakers",
        "running",
        "boots",
        "shoes",
        "sandals & slides",
        "slippers",
        "shoe care & acessories"
    ],
    "Accessories": [
        "bags",
        "belts",
        "hats",
        "jewellery",
        "socks",
        "scarves & gloves",
        "sunglasses",
        "wallets & keychains",
        "umbrellas",
        "watches"
    ],
    "Lifestyle": [
        "face & body",
        "glassware",
        "home decoration",
        "home fragrance",
        "lighting",
        "perfume & fragrance",
        "pet care",
        "publications",
        "soft furnishings",
        "storage & organisers",
        "tableware",
        "tech & audio"
    ]
}

product_descriptors = {
    "product_name": "The name of the item",
    "category": f"The category of the item. The options are: {category_options}",
    "sub-category": f"More specific categorisation under the category. The options for each category are represented as lists in the following dictionary: {sub_category_options}",
    "specific_category": "If there is even more specific categorisations not captured by sub-category. E.g., like backpack in the sub-category bags",
    "brand": "Brand Name",
    "fashion_season": "The season the item/collection came out. Mostly applicable to high end fashion. Only put it in the format SS (Spring/Summer) or AW (Autumn/Winter) and then the last two digits of the year like SS23 or AW24 ...",
    "primary_colour": f"The primary colour. The options are: {colour_options}",
    "secondary_colour": f"The secondary colour. The options are: {colour_options}",
    "pattern": f"The pattern on the clothing. The options are: {pattern_options}",
    "fabric": f"The material of the fabric. The options are: {fabric_options}",
    "price": "This is the price of the item",
    "price_category": f"The options are: [budget, mid-range, luxury]",
    "size": "sizing of the product",
    "fit": f"the way a piece of clothing fits on a person. The options are: {fit_options}",
    "gender_audience": f"The gender of the person the item is aimed at. The options are: {gender_audience_options}",
    "age_audience": f"The age of the person the item is aimed at. The options are: {gender_audience_options}",
    "season": f"What season is this aimed to be worn in. The options are: {season_options}",
    "occasion": f"The occasion the item is used for. The options are {occasion_option}",
    "sleeve_length": f"The length of the sleeve of a top. The options are {sleeve_length_options}",
    "neckline": f"The style of the neckline. The options are {neckline_options}",
    "country_of_manufacture": "Where the item was made.",
    "closure_type": f"Method of closing an item of clothing. The options are {closure_type_options}",
    "pocket_details": f"Details on the pockets of an item of clothing. The options are: {pocket_details_options}",
    "embellishments": f"Information on additional decorative elements. The options are: {embellishments_options}",
    "second_hand": "If the item is pre-owned. The options are: [true, false]",
    "sustainability": "If the item has some link to sustainability, like materials are organic or eco-friendly. Options are [true, false]",
    "additional_features": "Any other unique details like waterproofing, stretchability, or tech integration. that haven't been described by the attributes above.",
    "description": "A textual description that highlights special features or details. Be concise and focus on language that describes the item clearly for someone who has specific need",
    "key_words": "Distinct keywords that describe the item. Choose a maximum of 15 words/small phrases and separate each with a comma (,)"
}

Some ideas for building the most accurate retriever

- Using the feature dictionary as above. An LLM can use description analysis and image analysis to populate a dictionary for the item with words/phrases as each value using the above reference dictionary as an example.

- I could force the LLM to choose between specific words for each key-value pair and then use a simple search for specific options when finding a similar item.

- Or I could use short embeddings for description words and use a similarity check so that descriptors that are close to eachother come up in the search. E.g., very expensive, expensive, and cheap could be one of 5 categories but I would prefer that an item in the very expensive category is more "similar" than the "cheap" one. This is especially applicable to descriptors of style where the lines are slightly more blurred and natural language would be better than specific options.

- I may want to use an AI agent to look at the customer description to assign weights (or even hard stops) on features. For exmaple if a user asks for shorts specifically in the query then the agent sets a hard filter on the catgegory to match shorts. Or also if a person says something under £20 then we only want to show items under £20.

- When I retrieve some items I might want to use an AI agent to have a look at the recommendations and rank them in usefullness compared to the user query as an additional check at the end.

- I could include a 'addtional comments' bit of the description where I ask the AI to give a brief description of the style of the item and any other information that might be useful and use a vector store on that specific column in conjunction with the other search functions.

- An important aspect will be how to handle features of the item dictionary which are null. I won't be able to fill every feature with the user query. Maybe only comparing on a certain subset of features which can actually be determined from the query.

In [6]:
item = df.iloc[0]

image_urls_as_list = item['ImageUrls'].replace("[", "").replace("]", "").replace("'", "").split(", ")
item_dict = {
    "Product Name": item["ProductName"],
    "Brand": item["Brand"],
    "Price": item["Price"],
    "Category": item["Category"],
    "TextDescription": item["TextDescription"],
    "BulletPointDescription": item["BulletPointDescription"]
}

In [7]:
image_urls_as_list

['https://media.endclothing.com/media/f_auto,q_auto:eco,w_200/prodmedia/media/catalog/product/1/4/14-04-2023-JA_AAUSW0177FA01-BLK0001_1_1.jpg',
 'https://media.endclothing.com/media/f_auto,q_auto:eco,w_200/prodmedia/media/catalog/product/1/4/14-04-2023-JA_AAUSW0177FA01-BLK0001_7_1.jpg',
 'https://media.endclothing.com/media/f_auto,q_auto:eco,w_200/prodmedia/media/catalog/product/1/4/14-04-2023-JA_AAUSW0177FA01-BLK0001_8_1.jpg',
 'https://media.endclothing.com/media/f_auto,q_auto:eco,w_200/prodmedia/media/catalog/product/1/4/14-04-2023-JA_AAUSW0177FA01-BLK0001_m9_1.jpg',
 'https://media.endclothing.com/media/f_auto,q_auto:eco,w_200/prodmedia/media/catalog/product/1/4/14-04-2023-JA_AAUSW0177FA01-BLK0001_10_1.jpg']

In [ ]:
# Using OpenAIs gpt40-mini to extract features from images and the table data for an item
from openai import OpenAI

client = OpenAI(api_key="")

response = client.chat.completions.create(
    temperature=0,
    model="gpt-4o-mini",
    messages=[
        {
            "role": "system",
            "content": f"""You are tasked with extracting features for an item of clothing. 
            You will be provided with images of the item and some textual data from a database. 
            This is a dictionary of the features (keys) with a brief description of the feature (values): 
            \n\n{str(product_descriptors)}\n\n You should respond with only valid JSON where the keys are the keys from the dictionary just shown. If there is no valid value for a certain feature just put an empty string (''). For the values in the dictionary above some will have options that you need to stick to. Everything should be lower case."""
        },
        {
            "role": "user",
            "content": [
                {
                    "type": "text",
                    "text": f"This is the data on the item from the database: {str(item_dict)}",
                },
                {
                    "type": "image_url",
                    "image_url": {
                        "url": image_urls_as_list[0],
                    },
                }
            ],
        }
    ],
    max_tokens=500,
)

print(response.choices[0])

Choice(finish_reason='stop', index=0, logprobs=None, message=ChatCompletionMessage(content='```json\n{\n    "product_name": "1017 alyx 9sm circle logo popover hoodie",\n    "category": "clothing",\n    "sub-category": "hoodies & sweats",\n    "specific_category": "",\n    "brand": "1017-alyx-9sm",\n    "fashion_season": "",\n    "primary_colour": "black",\n    "secondary_colour": "",\n    "pattern": "graphic",\n    "fabric": "cotton",\n    "price": "325.0",\n    "price_category": "luxury",\n    "size": "",\n    "fit": "",\n    "gender_audience": "unisex",\n    "age_audience": "unisex",\n    "season": "all seasons",\n    "occasion": "casual",\n    "sleeve_length": "long",\n    "neckline": "drawstring hood",\n    "country_of_manufacture": "italy",\n    "closure_type": "drawstring",\n    "pocket_details": "kangaroo pocket",\n    "embellishments": "logo",\n    "second_hand": "false",\n    "sustainability": "false",\n    "additional_features": "heavyweight",\n    "description": "This heavyw

In [10]:
string_output_cheap = response.choices[0].message.content

In [11]:
string_output = response.choices[0].message.content

In [12]:
string_output

'```json\n{\n    "product_name": "1017 alyx 9sm circle logo popover hoodie",\n    "category": "clothing",\n    "sub-category": "hoodies & sweats",\n    "specific_category": "",\n    "brand": "1017-alyx-9sm",\n    "fashion_season": "",\n    "primary_colour": "black",\n    "secondary_colour": "",\n    "pattern": "graphic",\n    "fabric": "cotton",\n    "price": "325.0",\n    "price_category": "luxury",\n    "size": "",\n    "fit": "",\n    "gender_audience": "unisex",\n    "age_audience": "unisex",\n    "season": "all seasons",\n    "occasion": "casual",\n    "sleeve_length": "long",\n    "neckline": "drawstring hood",\n    "country_of_manufacture": "italy",\n    "closure_type": "drawstring",\n    "pocket_details": "kangaroo pocket",\n    "embellishments": "logo",\n    "second_hand": "false",\n    "sustainability": "false",\n    "additional_features": "heavyweight",\n    "description": "This heavyweight hoodie combines comfort and style, featuring a circle logo on the chest.",\n    "key_

In [13]:
import json

def transform_string_json_to_dict(string_output: str):

    lines = string_output.splitlines()
    if lines[0].startswith("```") and lines[-1].startswith("```"):
        json_str = "\n".join(lines[1:-1])
    else:
        json_str = string_output

    # Convert the JSON string to a Python dictionary
    data = json.loads(json_str)

    return data

In [14]:
dict_output = transform_string_json_to_dict(string_output)
dict_output_cheap = transform_string_json_to_dict(string_output_cheap)

In [15]:
dict_output

{'product_name': '1017 alyx 9sm circle logo popover hoodie',
 'category': 'clothing',
 'sub-category': 'hoodies & sweats',
 'specific_category': '',
 'brand': '1017-alyx-9sm',
 'fashion_season': '',
 'primary_colour': 'black',
 'secondary_colour': '',
 'pattern': 'graphic',
 'fabric': 'cotton',
 'price': '325.0',
 'price_category': 'luxury',
 'size': '',
 'fit': '',
 'gender_audience': 'unisex',
 'age_audience': 'unisex',
 'season': 'all seasons',
 'occasion': 'casual',
 'sleeve_length': 'long',
 'neckline': 'drawstring hood',
 'country_of_manufacture': 'italy',
 'closure_type': 'drawstring',
 'pocket_details': 'kangaroo pocket',
 'embellishments': 'logo',
 'second_hand': 'false',
 'sustainability': 'false',
 'additional_features': 'heavyweight',
 'description': 'This heavyweight hoodie combines comfort and style, featuring a circle logo on the chest.',
 'key_words': 'hoodie, circle logo, heavyweight, cotton, drawstring, kangaroo pocket, ribbed trims, casual, unisex, luxury, italy, ska

In [31]:
# Checking the output
output_dict_keys = set(dict_output_cheap.keys())
reference_dict_keys = set(product_descriptors.keys())

print(output_dict_keys.difference(reference_dict_keys))
print(reference_dict_keys.difference(output_dict_keys))

set()
set()


In [41]:
def compare_dicts(dict1, dict2):
    differences = {}
    # Create a set of all keys present in either dictionary
    all_keys = set(dict1.keys()) | set(dict2.keys())
    
    for key in all_keys:
        val1 = dict1.get(key)
        val2 = dict2.get(key)
        # If the values differ (or key is missing in one dictionary), record the difference
        if val1 != val2:
            differences[key] = [val1, val2]
    return differences

In [42]:
compare_dicts(dict_output, dict_output_cheap)

{}

Using 1 image is roughly 9.5k tokens or 0.15*(9500/1000000) = $0.0014
Using all 5 images is roughly 43k tokens or 0.15*(9500/1000000) = $0.00645

With 14912 items in my table it would cost me:
Using 1 image -> 14912*0.0014 = $20.88
Using all images (assuming a 5 image average) -> 14912*0.00645 = $96.18

For this 1 example using 1 image gave exactly the same output dictionary as using all the images so doesn't make sense to use all images (plus it  took 1s vs 5s so would increase the time taken by a lot)

In [ ]:
from openai import OpenAI

def get_product_descriptors(item: pd.Series, product_descriptors: dict):

    client = OpenAI(api_key="")

    image_urls_as_list = item['ImageUrls'].replace("[", "").replace("]", "").replace("'", "").split(", ")
    item_dict = {
        "Product Name": item["ProductName"],
        "Brand": item["Brand"],
        "Price": item["Price"],
        "Category": item["Category"],
        "TextDescription": item["TextDescription"],
        "BulletPointDescription": item["BulletPointDescription"]
    }

    if len(image_urls_as_list) > 0:

        response = client.chat.completions.create(
            model="gpt-4o-mini",
            messages=[
                {
                    "role": "system",
            "content": f"""You are tasked with extracting features for an item of clothing. 
            You will be provided with images of the item and some textual data from a database. 
            This is a dictionary of the features (keys) with a brief description of the feature (values): 
            \n\n{str(product_descriptors)}\n\n You should respond with only valid JSON where the keys are the keys from the dictionary just shown. If there is no valid value for a certain feature just put an empty string (''). For the values in the dictionary above some will have options that you need to stick to. Everything should be lower case."""
                },
                {
                    "role": "user",
                    "content": [
                        {
                            "type": "text",
                            "text": f"This is the data on the item from the database: {str(item_dict)}",
                        },
                        {
                            "type": "image_url",
                            "image_url": {
                                "url": image_urls_as_list[0],
                            },
                        }
                    ],
                }
            ],
            max_tokens=500,
        )
    else:
        response = client.chat.completions.create(
            model="gpt-4o-mini",
            messages=[
                {
                    "role": "system",
                    "content": f"""You are tasked with extracting features for an item of clothing. 
                    You will be provided with some textual data from a database. 
                    This is a dictionary of the features (keys) with a brief description of the feature (values): 
                    \n\n{str(product_descriptors)}\n\n You should respond with only valid JSON where the keys are the keys from the dictionary just shown. If there is no valid value for a certain feature just put an empty string ('')."""
                },
                {
                    "role": "user",
                    "content": [
                        {
                            "type": "text",
                            "text": f"This is the data on the item from the database: {str(item_dict)}",
                        }
                    ],
                }
            ],
            max_tokens=500,
        )

    string_output = response.choices[0].message.content

    product_dict = transform_string_json_to_dict(string_output)

    product_dict['image_urls'] = image_urls_as_list

    return product_dict

In [17]:
def generate_dataset_from_product_descriptors(df: pd.DataFrame, product_descriptors: dict):

    descriptor_list = []

    for i in range(len(df)):

        descriptor_list.append(get_product_descriptors(df.iloc[0], product_descriptors))

    return pd.DataFrame(descriptor_list)

In [18]:
descriptor_list = []
error_indices = []

for i in range(len(df)):
    try:
        descriptor_list.append(get_product_descriptors(df.iloc[i], product_descriptors))
    except:
        descriptor_list.append({})
        error_indices.append(i)

res_df = pd.DataFrame(descriptor_list)

In [19]:
res_df

product_name  category  \
0              1017 alyx 9sm circle logo popover hoodie  clothing   
1                      1017 alyx 9sm side zip camo pant  clothing   
2               1017 alyx 9sm x blackmeans denim jacket  clothing   
3      1017 alyx 9sm oversized camo nylon bomber jacket  clothing   
4               1017 alyx 9sm nylon logo varsity jacket  clothing   
...                                                 ...       ...   
14907                                               NaN       NaN   
14908                                               NaN       NaN   
14909                                               NaN       NaN   
14910                                               NaN       NaN   
14911                                               NaN       NaN   

           sub-category specific_category          brand fashion_season  \
0      hoodies & sweats                    1017-alyx-9sm                  
1              trousers                    1017-alyx-9sm                  
2       coats & jackets                    1017-alyx-9sm                  
3       coats & jackets                    1017-alyx-9sm                  
4       coats & jackets                    1017-alyx-9sm                  
...                 ...               ...            ...            ...   
14907               NaN               NaN            NaN            NaN   
14908               NaN               NaN            NaN            NaN   
14909               NaN               NaN            NaN            NaN   
14910               NaN               NaN            NaN            NaN   
14911               NaN               NaN            NaN            NaN   

      primary_colour secondary_colour  pattern     fabric  ... closure_type  \
0              black                   graphic     cotton  ...   drawstring   
1              green            brown     camo      nylon  ...          zip   
2              black                               cotton  ...       button   
3              green            brown     camo      nylon  ...  snap button   
4              black                     solid  polyester  ...       button   
...              ...              ...      ...        ...  ...          ...   
14907            NaN              NaN      NaN        NaN  ...          NaN   
14908            NaN              NaN      NaN        NaN  ...          NaN   
14909            NaN              NaN      NaN        NaN  ...          NaN   
14910            NaN              NaN      NaN        NaN  ...          NaN   
14911            NaN              NaN      NaN        NaN  ...          NaN   

        pocket_details embellishments second_hand sustainability  \
0      kangaroo pocket           logo       false          false   
1          back pocket                      false          false   
2         chest pocket                      false          false   
3          side pocket                      false          false   
4          side pocket           logo       false          false   
...                ...            ...         ...            ...   
14907              NaN            NaN         NaN            NaN   
14908              NaN            NaN         NaN            NaN   
14909              NaN            NaN         NaN            NaN   
14910              NaN            NaN         NaN            NaN   
14911              NaN            NaN         NaN            NaN   

                            additional_features  \
0                                                 
1      elasticated waist, full side zip closure   
2                 unique texture, sueded effect   
3                                                 
4         technical build guarantees durability   
...                                         ...   
14907                                       NaN   
14908                                       NaN   
14909                                       NaN   
14910                                

In [20]:
res_df.to_csv('ProductStructuredInfo2.csv')

In [1]:
error_indices_set = set(error_indices)

for indx in error_indices:
    try:
        descriptor_list[indx] = get_product_descriptors(df.iloc[indx], product_descriptors)
        error_indices_set.remove(indx)
    except:
        descriptor_list[indx] = {}

NameError: name 'error_indices' is not defined

In [73]:
res_df = pd.DataFrame(descriptor_list)

In [77]:
df_items = pd.read_csv('DataSets/ItemDataWithAiDescription.csv', index_col=0)